File: packages/contracts/crypto/conformance/kotlin/src/main/kotlin/com/uankit53/crewroll/crypto/Aad.kt
package com.uankit53.crewroll.crypto

import java.util.Locale

enum class MediaVariant(val code: Int) {
    PREVIEW(1),
    ORIGINAL(2),
}

object CrewRollAad {
    fun media(
        tripId: String,
        assetId: String,
        variant: MediaVariant,
        keyEpoch: Long = 1,
        formatVersion: Long = 1,
    ): ByteArray {
        if (keyEpoch != 1L || formatVersion != 1L) {
            throw CryptoReadException(
                CryptoFailure.SEMANTIC_CONTEXT,
                "media AAD epoch and format must both be 1",
            )
        }
        return BinaryWriter(95).apply {
            writeAscii("CRROLL-AAD-V1\u0000")
            writeAscii(canonicalUuid(tripId))
            writeAscii(canonicalUuid(assetId))
            writeU8(variant.code)
            writeU32(keyEpoch)
            writeU32(formatVersion)
        }.toByteArray().also {
            check(it.size == 95)
        }
    }

    fun manifest(
        tripId: String,
        assetId: String,
        keyEpoch: Long = 1,
        formatVersion: Long = 1,
    ): ByteArray {
        if (keyEpoch != 1L || formatVersion != 1L) {
            throw CryptoReadException(
                CryptoFailure.SEMANTIC_CONTEXT,
                "manifest AAD epoch and format must both be 1",
            )
        }
        return BinaryWriter(94).apply {
            writeAscii("CRROLL-MAN-V1\u0000")
            writeAscii(canonicalUuid(tripId))
            writeAscii(canonicalUuid(assetId))
            writeU32(keyEpoch)
            writeU32(formatVersion)
        }.toByteArray().also {
            check(it.size == 94)
        }
    }
}

private val canonicalUuidPattern = Regex(
    "^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$",
)

internal fun canonicalUuid(value: String): String {
    if (!canonicalUuidPattern.matches(value)) {
        throw CryptoReadException(
            CryptoFailure.STRUCTURE,
            "UUID must have canonical shape",
        )
    }
    return value.lowercase(Locale.ROOT)
}

internal fun uuidBytes(value: String): ByteArray {
    val compact = canonicalUuid(value).replace("-", "")
    return ByteArray(16) { index ->
        compact.substring(index * 2, index * 2 + 2).toInt(16).toByte()
    }
}

File: modules/crewroll-transfer/ios/IdentityKeys/Sources/CrewRollNativeKeys/MediaCrypto/AAD.swift
import Foundation

public enum MediaVariant: UInt8, CaseIterable {
    case preview = 1
    case original = 2
}

public enum CrewRollAAD {
    public static func media(
        tripID: String,
        assetID: String,
        variant: MediaVariant,
        keyEpoch: UInt32 = 1,
        formatVersion: UInt32 = 1
    ) throws -> Bytes {
        guard keyEpoch == 1, formatVersion == 1 else {
            throw CryptoReadError.semanticContext(
                "media AAD epoch and format must both be 1"
            )
        }
        var writer = ByteWriter()
        try writer.writeASCII("CRROLL-AAD-V1\0")
        try writer.writeASCII(canonicalUUID(tripID))
        try writer.writeASCII(canonicalUUID(assetID))
        writer.write(variant.rawValue)
        writer.write(keyEpoch)
        writer.write(formatVersion)
        guard writer.bytes.count == 95 else {
            throw CryptoReadError.structure("media AAD must be exactly 95 bytes")
        }
        return writer.bytes
    }

    public static func manifest(
        tripID: String,
        assetID: String,
        keyEpoch: UInt32 = 1,
        formatVersion: UInt32 = 1
    ) throws -> Bytes {
        guard keyEpoch == 1, formatVersion == 1 else {
            throw CryptoReadError.semanticContext(
                "manifest AAD epoch and format must both be 1"
            )
        }
        var writer = ByteWriter()
        try writer.writeASCII("CRROLL-MAN-V1\0")
        try writer.writeASCII(canonicalUUID(tripID))
        try writer.writeASCII(canonicalUUID(assetID))
        writer.write(keyEpoch)
        writer.write(formatVersion)
        guard writer.bytes.count == 94 else {
            throw CryptoReadError.structure(
                "manifest AAD must be exactly 94 bytes"
            )
        }
        return writer.bytes
    }
}

func canonicalUUID(_ value: String) throws -> String {
    let segments = value.split(separator: "-", omittingEmptySubsequences: false)
    let widths = [8, 4, 4, 4, 12]
    guard segments.count == widths.count else {
        throw CryptoReadError.structure("UUID must have canonical shape")
    }
    for (segment, width) in zip(segments, widths) {
        guard segment.count == width,
              segment.utf8.allSatisfy({ byte in
                  (48 ... 57).contains(byte)
                      || (65 ... 70).contains(byte)
                      || (97 ... 102).contains(byte)
              })
        else {
            throw CryptoReadError.structure("UUID must have canonical shape")
        }
    }
    return value.lowercased()
}

func uuidBytes(_ value: String) throws -> Bytes {
    let canonical = try canonicalUUID(value)
    let compact = canonical.replacingOccurrences(of: "-", with: "")
    var result: Bytes = []
    result.reserveCapacity(16)
    var position = compact.startIndex
    for _ in 0 ..< 16 {
        let next = compact.index(position, offsetBy: 2)
        guard let byte = UInt8(compact[position ..< next], radix: 16) else {
            throw CryptoReadError.structure("UUID contains invalid hex")
        }
        result.append(byte)
        position = next
    }
    return result
}
